### PR TITLE
Add windows tests

### PR DESCRIPTION
**Pull request by roidelapluie**
Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -55,8 +55,14 @@ jobs:
     working_directory: /go/src/github.com/prometheus/prometheus
     steps:
     - checkout
-    # TSDB is where the most risk is Windows wise, so only test there for now.
-    - run: go test ./tsdb/...
+    - run:
+        shell: bash
+        command: |
+          (cd web/ui && GOOS= GOARCH= go generate -mod=vendor)
+          go test -mod=vendor -test.v `go list ./...|grep -v github.com/prometheus/prometheus/discovery | grep -xv github.com/prometheus/prometheus/config | grep -xv github.com/prometheus/prometheus/web`
```
Comment (brian-brazil) [inline .circleci/config.yml]: You can use -e with grep to have multiple patterns

**Comment by roidelapluie** [comment]
Rebased on top of #6658

File: cmd/prometheus/main_unix_test.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	"os/exec"
+	"testing"
+	"time"
+)
+
+func stopProcess(p *os.Process) {
+	p.Signal(os.Interrupt)
+}
+
+// As soon as prometheus starts responding to http request should be able to accept Interrupt signals for a graceful shutdown.
```
Comment (brian-brazil) [inline cmd/prometheus/main_unix_test.go]: requests it should

File: cmd/prometheus/main_unix_test.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	go func() {
+		done <- prom.Wait()
+	}()
+
+	var startedOk bool
+	var stoppedErr error
+
+Loop:
+	for x := 0; x < 10; x++ {
+		// error=nil means prometheus has started so can send the interrupt signal and wait for the grace shutdown.
```
Comment (brian-brazil) [inline cmd/prometheus/main_unix_test.go]: so we can

graceful

File: cmd/prometheus/main_windows_test.go [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+// +build windows
+
+package main
+
+import (
+	"os"
+)
+
+func stopProcess(p *os.Process) {
+	p.Kill()
```
Comment (brian-brazil) [inline cmd/prometheus/main_windows_test.go]: Why do we not stop all OSes the same way?

File: discovery/manager_windows_test.go [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// +build windows
+
+package discovery
+
+import "time"
+
+var timescale = 10 * time.Millisecond
```
Comment (brian-brazil) [inline discovery/manager_windows_test.go]: Please don't indirect basic constants like this, and if one OS needs more time give it to all of them.

File: util/stats/stats_unix_test.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// +build !windows
+
+package stats
+
+import "testing"
+
+// TestTimerGroup does not work on Windows since its timer is not accurate
```
Comment (brian-brazil) [inline util/stats/stats_unix_test.go]: That doesn't sound right, Windows has high resolution timers too. This sounds like a bug in the timer code.

File: cmd/prometheus/main_windows_test.go (hunk 4 above)
Comment (roidelapluie) [reply cmd/prometheus/main_windows_test.go]: It is preferrable to stop with Interrupt as it gives Prometheus the ability to cleanly shutdown.

However Interrupt is not supported in Windows.

File: util/stats/stats_unix_test.go (hunk 6 above)
Comment (roidelapluie) [reply util/stats/stats_unix_test.go]: Time.Time() returns the same if call consecutively on Win.

Added  Sleep()

File: cmd/prometheus/main_windows_test.go (hunk 4 above)
Comment (brian-brazil) [reply cmd/prometheus/main_windows_test.go]: KILL is supported on Unix.

File: util/stats/stats_unix_test.go (hunk 6 above)
Comment (brian-brazil) [reply util/stats/stats_unix_test.go]: What version of Windows? It really sounds like we're using the wrong timing API here.

File: cmd/prometheus/main_windows_test.go (hunk 4 above)
Comment (roidelapluie) [reply cmd/prometheus/main_windows_test.go]: does it trigger a graceful shutdown of TSDB?

Comment (brian-brazil) [reply cmd/prometheus/main_windows_test.go]: No, but you have that exact problem on Windows anyway.

File: util/stats/timer.go [hunk 7]
```diff
@@ -95,6 +95,7 @@ func (t Timers) Swap(i, j int) {
 }
 
 func (s byCreationTimeSorter) Less(i, j int) bool {
+	fmt.Printf("i: %v / j: %v\n", s.Timers[i].created, s.Timers[j])
```
Comment (brian-brazil) [inline util/stats/timer.go]: Debug code?

Comment (roidelapluie) [reply util/stats/timer.go]: yes to debug the time.Time issue

Comment (roidelapluie) [reply util/stats/timer.go]: here is a version that does not rely on time.Time

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**